### PR TITLE
Add rules for spaces inside function calls [WIP]

### PR DESCRIPTION
**Pull request by texel**
Speculative work to address issue #1916

### CODE REVIEW

File: lib/rules/disallow-spaces-inside-function-calls.js [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+ *
+ * #### Example
+ *
+ * ```js
+ * "disallowSpacesInsideFunctionCalls": true
+ *
+ * // or
+ *
+ * "disallowSpacesInsideFunctionCalls": {
+ *     "allExcept": [ "'", "\'" ]
```
Comment (zxqfox) [inline lib/rules/disallow-spaces-inside-function-calls.js]: `"'" === "\'"` right?

File: lib/rules/disallow-spaces-inside-function-calls.js [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+ * - true
+ * - `Object`:
+ *      - "only": An array containing tokens to disallow.
+ *      - "onlySingleArgument": true
+ *      - "onlyMultipleArguments": true
+ *
+ * #### Example
+ *
+ * ```js
+ * "disallowSpacesInsideFunctionCalls": true
```
Comment (zxqfox) [inline lib/rules/disallow-spaces-inside-function-calls.js]: What is the default?

File: lib/rules/disallow-spaces-inside-function-calls.js [hunk 3]
```diff
@@ -0,0 +1,143 @@
+/**
+ * Disallows space after opening and before closing function call parentheses.
+ *
+ * Types: `Boolean` or `Object`
+ *
+ * Values:
+ * - true
+ * - `Object`:
+ *      - "only": An array containing tokens to disallow.
```
Comment (zxqfox) [inline lib/rules/disallow-spaces-inside-function-calls.js]: I'd say we don't need too much options. Can we left just 2 params: `allExcept` with tokens array and, let's say, `arity` with 2 variants: single, multiple or even a Number.

File: lib/rules/disallow-spaces-inside-function-calls.js [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+
+var assert = require('assert');
+
+module.exports = function() {};
+
+module.exports.prototype = {
+    configure: function(value) {
+        var isObject = typeof value === 'object';
+
+        var error = this.getOptionName() + ' rule requires string value true or object';
```
Comment (zxqfox) [inline lib/rules/disallow-spaces-inside-function-calls.js]: "string value true"? nooo.

File: lib/rules/disallow-spaces-inside-function-calls.js [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+        var validFlags = [
+            'only',
+            'onlySingleArgument',
+            'onlyMultipleArguments'
+        ];
+
+        if (isObject) {
+            var hasValidFlags;
+
+            validFlags.forEach(function(flag) {
```
Comment (zxqfox) [inline lib/rules/disallow-spaces-inside-function-calls.js]: Better to check keys to exist in validFlags. Like `assert(Object.keys(value).every(function (key) { return validFlags.indexOf(key) !== -1; }))`

File: lib/rules/disallow-spaces-inside-function-calls.js [hunk 6]
```diff
@@ -0,0 +1,143 @@
...
+                hasValidFlags = hasValidFlags || flag in value;
+            });
+
+            assert(hasValidFlags, error);
+        } else {
+            assert(value === true, error);
```
Comment (zxqfox) [inline lib/rules/disallow-spaces-inside-function-calls.js]: Why not
```
if (!isObject) {
  assert(value === true, error);
  return;
}

// do the rest
```
? ;-) We usually like early returns.

File: lib/rules/disallow-spaces-inside-function-calls.js [hunk 7]
```diff
@@ -0,0 +1,143 @@
...
+ * ```
+ */
+
+var assert = require('assert');
+
+module.exports = function() {};
+
+module.exports.prototype = {
+    configure: function(value) {
+        var isObject = typeof value === 'object';
```
Comment (zxqfox) [inline lib/rules/disallow-spaces-inside-function-calls.js]: Better to check for isArray also.

File: lib/rules/disallow-spaces-inside-function-calls.js [hunk 8]
```diff
@@ -0,0 +1,143 @@
...
+                    token: argLastToken,
+                    nextToken: closeParens,
+                    message: 'Space before closing parenthesis'
+                });
+            } else {
+                var callToken = file.getTokenByRangeStart(functionCall.range[0]);
+                var openParen = file.getNextToken(callToken);
+                var closeParen = file.getNextToken(openParen);
+
+                errors.assert.noWhitespaceBetween({
```
Comment (zxqfox) [inline lib/rules/disallow-spaces-inside-function-calls.js]: same here — better to move to the top of the function, right after single/multiple checks (or even before them).

File: lib/rules/disallow-spaces-inside-function-calls.js [hunk 9]
```diff
@@ -0,0 +1,143 @@
...
+
+    check: function(file, errors) {
+        var only = this._only;
+        var onlySingleArgument = this._onlySingleArgument;
+        var onlyMultipleArguments = this._onlyMultipleArguments;
+
+        file.iterateNodesByType('CallExpression', function(functionCall) {
+            var functionArgs = functionCall.arguments;
+
+            if (onlySingleArgument && functionArgs.length !== 1) {
```
Comment (zxqfox) [inline lib/rules/disallow-spaces-inside-function-calls.js]: What about noArgs case? Is it fine here to have a space between parens?

File: lib/rules/disallow-spaces-inside-function-calls.js [hunk 10]
```diff
@@ -0,0 +1,143 @@
...
+            } else if (onlyMultipleArguments && functionArgs.length === 1) {
+                return;
+            }
+
+            if (functionArgs.length) {
+                var firstArg = functionArgs[0];
+                var lastArg = functionArgs[functionArgs.length - 1];
+                var argFirstToken = file.getTokenByRangeStart(firstArg.range[0]);
+                var argLastToken = file.getTokenByRangeEnd(lastArg.range[1]);
+                var openParens = file.getPrevToken(argFirstToken);
```
Comment (zxqfox) [inline lib/rules/disallow-spaces-inside-function-calls.js]: openParens or openParen?

**Comment by gibson042** [comment]
I'm really sorry to jump in so late on this, but I have some major reservations.

# The Name
Consistency with #1254 would use **SpacesInsideParenthesizedArguments** if optimizing for alphabetical grouping, or [**SpacesInsideCall**(**Expression**)?**Arguments**](https://github.com/jscs-dev/node-jscs/pull/1254#issuecomment-133056398) if optimizing for unambiguous comprehension. Which way to go remains an open question, since it was asked in https://github.com/jscs-dev/node-jscs/pull/1254#issuecomment-133107488 but never resolved. But **SpacesInsideFunctionCalls** is in my opinion _far_ too similar to the existing **[SpacesInCallExpression](http://jscs.info/rule/disallowSpacesInCallExpression)** rules (`foo‸(a, b, …)`).

# The Implementation
1acc0c0b4c6afc1d2fab9257712977b76e887450 introduced `TokenCategorizer.categorizeOpenParen` and `TokenCategorizer.categorizeCloseParen` for just such purposes, and all parentheses rules should use them, lest we miss edge cases in unpredictable and inconsistent ways. For example,  `invocation( (parenthesizedExpression) )` has an open parenthesis preceding `file.getTokenByRangeStart(firstArg.range[0])` that is **not** wrapping an argument list, which will trip up the code in this PR.

**Comment by markelog** [comment]
@texel Friendly ping :-)

**Comment by markelog** [comment]
Okay, i'm closing it for now, ping us up if you want to continue work on this

**Pull request closed without merging**